Fe 26 hashtags component

dominikapap opened the pull request.

dominikapap committed:
style: modal content

added styles to modal content

dominikapap committed:
Merge branch 'develop' into FE-31-after-form-view-update

dominikapap committed:
style: form after view

changed dialog bg and container colors, overwriten material variables

dominikapap committed:
feat: routing for hashtags

added basic routing for event form

dominikapap committed:
feat: errors

added mat-error and validator for max-length

dominikapap committed:
refactor: code review changes

added ngSrc directive to image, moved icons url to a variable

dominikapap committed:
Merge branch 'develop' into FE-26-hashtags-component

dominikapap committed:
Merge branch 'develop' into FE-31-after-form-view-update

dominikapap committed:
refactor: merging conflicts

resolved conflicts

dominikapap committed:
Merge branch 'develop' into FE-31-after-form-view-update

dominikapap committed:
fix: unnecessary code

removed unused code

dominikapap committed:
refactor: routing

changed routing paths

dominikapap committed:
Merge branch 'FE-31-after-form-view-update' into FE-26-hashtags-component

dominikapap committed:
fix: merge issues

fixed merge issues

dominikapap committed:
refactor: merge issues

resolved more merging issues

dominikapap committed:
feat: hashtags component

added highlight text

dominikapap committed:
feat: snackbarService

moved snackBar logic to a service

dominikapap committed:
feat: snackbarService

moved snackBar logic to a service

dominikapap committed:
refactor: removed unused import

removed import of snackbarmodule from events module

dominikapap committed:
refactor: moved code to a function

moved code used for info about hashtag already added to a function

dominikapap committed:
feat: hashtag component

added styles for autocomplete focus and options for better UX

dominikapap committed:
refactor: before CR

moved a number to a variable

dominikapap committed:
refactor: snackbar method

moved code to a separate method

dominikapap committed:
fix: adding hashtag with autocomplete panel opened

changed option panelOpen to activeOption

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags.component.ts:
```diff
@@ -0,0 +1,152 @@
+    'Urodziny',
+    'BaaaaaaaaaardzoDłuuuuuugiHashtagAwogóletoooooooooo',
+    '한국말',
+  ];
```
Ja bym przeniósł mockowane dane do oddzielnego pliku. Mniej czasu zajmie usunięcie tego i HastagsComponent będzie bardziej przejrzysty.

dominikapap replied:
poprawione

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags.component.ts:
```diff
@@ -0,0 +1,152 @@
+    if (this.addedHashtags.length >= this.maxHashtagsCount) {
+      this.maxHashtagCountInfo();
+    } else {
+      const value = (event.value || '').trim().slice(0, this.maxChars);
```
może  nazwać to bardziej deskryptywnie
```
      const rawHashtag = (event.value || '').trim().slice(0, this.maxChars);
```
albo
```
      const hastagStr = (event.value || '').trim().slice(0, this.maxChars);
```

dominikapap replied:
zmienione na `hashtagStr`

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags.component.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  remove(hashtag: string): void {
+    const index = this.addedHashtags.indexOf(hashtag);
```
mało deskryptywnie, podobnie jak wyżej.

dominikapap replied:
poprawione

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags.component.ts:
```diff
@@ -0,0 +1,152 @@
+      if (value && !this.addedHashtags.includes(value) && !this.auto.activeOption) {
+        this.addedHashtags.push(value);
+      } else if (this.addedHashtags.includes(value)) {
+        this.hashtagAlreadyAddedInfo(value);
```
dobrze gdyby metoda ```hashtagAlreadyAddedInfo(value)``` była bezpośrednio pod metodą w której jest po raz pierwszy użyta, wtedy lepiej czyta się kod, nie trzeba scrollować po pliku w jej poszukiwaniu :)

dominikapap replied:
` hashtagAlreadyAddedInfo(value)` jest metodą prywatną, nie chciałam mieszać publicznych z prywatnymi więc jest na końcu, ale w VSC można się przenieść od razu do definicji metody jeśli się na nią kliknie trzymając przycisk Ctrl

dominikapap committed:
refactor: hashtags service

dominikapap committed:
Merge branch 'FE-26-hashtags-component' into develop

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,119 @@
+    'BaaaaaaaaaardzoDłuuuuuugiHashtagAwogóletoooooooooo',
+  ];
+
+  addHashtag(
```
Metoda ``addHashtag`` jest wewnętrz klasy Hastag, nazwa klasy wskazuje do czego się odnosi kod który jest wewnątrz, więc proponuję nazwać tę metodę ``add``

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  addHashtagFromAutocomplete(
```
jw

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  editHashtag(hashtag: string, event: MatChipEditedEvent, auto: MatAutocompleteTrigger) {
```
jw.

LukaszStaniszewski commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  removeAddedHashtag(hashtag: string, auto: MatAutocompleteTrigger): void {
```
jw

dominikapap committed:
refactor: hahshtag service

changed method names to be shorter

dominikapap committed:
fix: method name

renamed method to be consistent with others

abdevworks commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,119 @@
+})
+export class HashtagsService {
+  private snackBarService = inject(SnackBarService);
+  maxChars = 50;
+  maxHashtagsCount = 100;
```
Jako, że to są stałe używane jedynie w serwisie to powinny być prywatne oraz readonly. Dodatkowo można zmienić nazwę na MAX_CHARS oraz MAX_HASHTAGS_COUNT

dominikapap replied:
poprawione

dominikapap committed:
refactor: theme path

changed theme path for events module

dominikapap committed:
fix: autocomplete open

fixed opening autocomplete at the beggining

dominikapap committed:
refactor: private methods

changed methods in service to be private

abdevworks commented on src/app/features/event/event-form/hashtags/hashtag.component.html:
```diff
@@ -0,0 +1,53 @@
+        </mat-chip-row></mat-chip-grid
+    >
+    <mat-form-field appearance="outline"
+        ><mat-label>Hashtags</mat-label>
```
Na makiecie było pewnie po angielsku, ale nasza aplikacja jest jednak po polsku więc powinno być 'Hashtagi'

dominikapap replied:
poprawione

abdevworks commented on src/app/features/event/event-form/hashtags/hashtags.component.ts:
```diff
@@ -0,0 +1,81 @@
+} from '@angular/material/autocomplete';
+import { MatChipEditedEvent, MatChipInputEvent } from '@angular/material/chips';
+import { Observable } from 'rxjs';
+import { map, skip, startWith } from 'rxjs/operators';
```
Operatory skip oraz startWith nie są nigdzie używane, więc można je usunąć.

dominikapap replied:
poprawione

abdevworks commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,129 @@
+  MatAutocompleteTrigger,
+} from '@angular/material/autocomplete';
+import { MatChipInputEvent, MatChipEditedEvent } from '@angular/material/chips';
+import { SnackBarService } from '@shared/services/snack-bar.service';
```
Snack-bar service powinien być dodany do index.ts, aby można byłouprościć import do '@shared/services'

dominikapap replied:
słuszna uwaga, poprawiona

abdevworks commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,129 @@
+})
+export class HashtagsService {
+  private snackBarService = inject(SnackBarService);
+  private maxChars = 50;
+  private maxHashtagCount = 100;
```
Zgodnie z rozmową, zmienić, aby było również  readonly

dominikapap replied:
poprawione

abdevworks commented on src/app/shared/pipes/hihglightPipe.ts:
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-useless-escape */
```
Literówka. Zmienić nazwę pliku na 'highlightPipe'

dominikapap replied:
poprawione

abdevworks commented on src/app/features/event/event.module.ts:
```diff
@@ -1,14 +1,58 @@
+import { MatAutocompleteModule } from '@angular/material/autocomplete';
 import { CardComponent } from '.';
+import { EventsThemeComponent } from './event-form/event-form-theme.component';
+import { HighlightPipe } from '../../shared/pipes/hihglightPipe';
```
Czy, Pipes też nie powinny mieć swojego pliku index.ts?

dominikapap replied:
poprawione

abdevworks commented on src/app/features/event/event.module.ts:
```diff
@@ -1,14 +1,58 @@
+import { CommonModule } from '@angular/common';
 import { PowerIconComponent } from '@shared/ui';
+import { RouterModule } from '@angular/router';
+import { EventFormComponent } from './event-form/event-form.component';
```
EventFormComponent oraz HashtagsComponent można dodać do index.ts. Wtedy zamiast pełnej ścieżki można je ładnie zaimportować jako: import { CardComponent, EventFormComponent, HashtagsComponent } from '.';

dominikapap replied:
poprawione

dominikapap committed:
refactor: cr comments

added readonly to some private variables, added export to index.ts in pipes and events

abdevworks reviewed (approved):
Sprawdzone. Po wprowadzanych zmianach dla mnie wygląda ok.

akademiaEnvelo commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,129 @@
+  private readonly MAX_CHARS = 50;
+  private readonly MAX_HASHTAG_COUNT = 100;
+  private addedHashtags: string[] = [];
+  private allHashtags: string[] = [
+    'Impreza',
+    'Piwo',
+    'Piłkarzyki',
+    'Pizza',
+    'Pizza 2',
+    'Pizza 3',
+    'Pizza 4',
+    'Pizza 5',
+    'Karaoke',
+    'Bilard',
+    'Urodziny',
+    'BaaaaaaaaaardzoDłuuuuuugiHashtagAwogóletoooooooooo',
+  ];
```
co to za cudeńko?

dominikapap replied:
Zmockowane dane do autocomplete, zgodnie z taskiem na ten moment używamy danych zmockowanych w komponencie/serwisie. Jeśli tester będzie chciał to dodam więcej opcji :)

akademiaEnvelo commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,129 @@
+    hashtagInput: ElementRef<HTMLInputElement>,
+    hashtagCtrl: FormControl
+  ): void {
+    if (this.addedHashtags.length >= this.MAX_HASHTAG_COUNT) {
+      this.maxHashtagCountInfo();
+    } else {
+      const hashtagStr = event.option.viewValue;
+      if (this.addedHashtags.includes(hashtagStr)) {
+        this.hashtagAlreadyAddedInfo(hashtagStr);
+      } else {
+        hashtagInput.nativeElement.value = '';
+        hashtagCtrl.setValue('');
+        this.addedHashtags.push(hashtagStr);
+      }
+    }
```
by zmniejszyć ilośc zagnieżdzeń warto zastosować guard clause https://learningactors.com/javascript-guard-clauses-how-you-can-refactor-conditional-logic/

dominikapap replied:
zrobiłam refactor kodu, prośba o sprawdzenie ponowna czy nowa wersja jest akceptowalna

akademiaEnvelo commented on src/app/shared/pipes/highlightPipe.ts:
```diff
@@ -0,0 +1,19 @@
+    if (!match) {
+      return text;
+    }
+    return search ? text.replace(regex, match => `<b>${match}</b>`) : text;
```
zwracamy tutaj element html co w pipe jest taką średnio przyjętą praktyką, może po prostu warto pomyśleć o dyrektywie, ktora bedzie nadawala na dany element odpowiednią klase css?

Pipe co do zasady ma zmieniać nam wartość, jest swego rodzaju mapperem, a dyrektywa wplywa tez na kształt wartosci, działania i wyglądu właśnie

dominikapap replied:
zmienione na dyrektywę

dominikapap committed:
refactor: checking errors

moved checking errors to a separate method

dominikapap committed:
Merge branch 'develop' of https://github.com/akademia-envelo-3/MELO-front into develop

dominikapap committed:
feat: highlight directive

added hightlight directive

dominikapap committed:
refactor: highlight directive

refactor code in directive

dominikapap committed:
Merge branch 'develop' of https://github.com/akademia-envelo-3/MELO-front into develop

dominikapap committed:
fix: merge issues

removed faulty added files

dominikapap committed:
Merge branch 'develop' into FE-26-hashtags-component

dominikapap committed:
refactor: after merging develop

adjusted code to changes from develop branch

dominikapap committed:
fix: highlight directive

fixed directive to work on mat-option element

dominikapap committed:
fix: snackbar display

fixed snackbar not showing after adding navbar

dominikapap committed:
refactor:  package.json

removed changes from package.json

dominikapap committed:
fix: unnecessary files

removed files

dominikapap committed:
fix: unit name validator

removed unused file

akademiaEnvelo commented on src/app/features/event/event-form/hashtags/hashtags.service.ts:
```diff
@@ -0,0 +1,153 @@
+    } else if (this.alreadyAdded(hashtagStr)) {
+      this.hashtagAlreadyAddedInfo(hashtagStr);
+      return false;
+    } else if (auto && auto.activeOption) {
+      return false;
+    } else if (!hashtagStr) {
+      return false;
```
te dwa przypadki możemy połączyć

dominikapap replied:
poprawione

akademiaEnvelo commented on src/app/shared/directives/highlight.directive.ts:
```diff
@@ -0,0 +1,23 @@
+  selector: 'mat-option[appHighlight]',
+  standalone: true,
+})
+export class HighlightDirective implements OnChanges {
```
możemy w sumie wskazać w nazwie że dotyczy to MatOption

dominikapap replied:
poprawione

akademiaEnvelo commented on src/app/shared/directives/remove-spaces.directive.ts:
```diff
@@ -0,0 +1,14 @@
+
+  @HostListener('input', ['$event']) onInputChange() {
+    const initalValue = this.el.nativeElement.value;
+    this.el.nativeElement.value = initalValue.replace(/\s/g, '');
```
użyjmy `replaceAll`, wtedy pozbędziemy się tego cudownego regexa

dominikapap replied:
poprawione

akademiaEnvelo commented on src/app/shared/services/snack-bar.service.ts:
```diff
@@ -0,0 +1,31 @@
+  private horizontalPosition: MatSnackBarHorizontalPosition = 'center';
+  private verticalPosition: MatSnackBarVerticalPosition = 'top';
+
+  openSnackBar(
+    message: string,
+    duration = 3000,
+    action = 'Zamknij',
+    horizontalPosition = this.horizontalPosition,
+    verticalPosition = this.verticalPosition,
+    panelClass = ['red-snackbar']
```
jak mamy tak dużo parametrów funkcji to zróbmy z nich obiekt konfiguracyjny tak by wywoływać tę metodę w taki sposób, gdzie config jest oczywiście pcjonalny bo ma domyślne wartości `openSnackBar(message, config)`

dominikapap replied:
poprawione

dominikapap committed:
refactor: code readability

refactored code in highlight and remove spaces directives, snack bar and hashtag services

dominikapap committed:
refactor: snackbar service

changed method name from openSnackBar to open
